pkg/backend/openshiftcluster.go

```diff
@@ -239,34 +267,39 @@ func (ocb *openShiftClusterBackend) updateAsyncOperation(ctx context.Context, lo
 		})
 
 		if err != nil {
-			return err
+			return *operationResult, err
 		}
 	}
 
-	return nil
+	return *operationResult, nil
 }
 
 func (ocb *openShiftClusterBackend) endLease(ctx context.Context, log *logrus.Entry, stop func(), doc *api.OpenShiftClusterDocument, provisioningState api.ProvisioningState, backendErr error) error {
 	var adminUpdateError *string
 	var failedProvisioningState api.ProvisioningState
+	initialProvisioningState := doc.OpenShiftCluster.Properties.ProvisioningState
 
-	if doc.OpenShiftCluster.Properties.ProvisioningState != api.ProvisioningStateAdminUpdating &&
+	if initialProvisioningState != api.ProvisioningStateAdminUpdating &&
 		provisioningState == api.ProvisioningStateFailed {
-		failedProvisioningState = doc.OpenShiftCluster.Properties.ProvisioningState
+		failedProvisioningState = initialProvisioningState
 	}
 
 	// If cluster is in the non-terminal state we are still in the same
 	// operational context and AsyncOperation should not be updated.
 	if provisioningState.IsTerminal() {
-		err := ocb.updateAsyncOperation(ctx, log, doc.AsyncOperationID, doc.OpenShiftCluster, provisioningState, failedProvisioningState, backendErr)
+		operationResult, err := ocb.updateAsyncOperation(ctx, log, doc.AsyncOperationID, doc.OpenShiftCluster, provisioningState, failedProvisioningState, backendErr)
 		if err != nil {
 			return err
 		}
-
+		log.WithFields(logrus.Fields{
+			"resultType":    operationResult.resultType,
+			"operationType": initialProvisioningState.String(),
+			"errorDetails":  operationResult.errorDetails,
+		}).Print("Long running operation result")
```

Comment:
Can we do something like this?

1. Keep signature of `updateAsyncOperation` unchanged. In this iteration of the PR this method returns `operationResult`. And one might think that we always call `updateAsyncOperation` when an async operation is complete (which is not the case). Not returning `operationResult` will help to avoid confusion.
2. Drop `operationResult` struct altogether and incapsulate result logging in a function (see an example below).
    * We might want to keep `resultType` constatns somewhere, but we should probably find it a new home. Maybe somewhere under `github.com/Azure/ARO-RP/pkg/util/log`?

```suggestion
		ocb.asyncOperationResultLog(log, initialProvisioningState, backendErr)
```

P.S.: Better name suggestions for `asyncOperationResultLog` are welcome :)

And `asyncOperationResultLog` will be something like this:

```go
func (ocb *openShiftClusterBackend) asyncOperationResultLog(log *logrus.Entry, initialProvisioningState api.ProvisioningState, backendErr error) {
	log = log.WithFields(logrus.Fields{
		"resultType":    successResultType,
		"operationType": initialProvisioningState.String(),
	})

	if backendErr == nil {
		log.Info("operation successfully completed")
		return
	}

	_, ok := backendErr.(*api.CloudError)
	if ok {
		log = log.WithField("resultType", userErrorResultType)
	} else {
		log = log.WithField("resultType", serverErrorResultType)
	}

	log = log.WithField("errorDetails", backendErr.Error())
	log.Info("operation failed")
	// Or maybe just add an error into `msg` like this,
	// instead of 2 lines of code above?
	log.Error(backendErr)
}

```